pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -184,120 +190,193 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
-			}).ViaFieldIndex("volumeMounts", i))
+			}).ViaIndex(i))
 		}
 		seen.Insert(vm.Name)
 
 		if vm.MountPath == "" {
-			errs = errs.Also(apis.ErrMissingField("mountPath").ViaFieldIndex("volumeMounts", i))
+			errs = errs.Also(apis.ErrMissingField("mountPath").ViaIndex(i))
 		} else if reservedPaths.Has(filepath.Clean(vm.MountPath)) {
 			errs = errs.Also((&apis.FieldError{
 				Message: fmt.Sprintf("mountPath %q is a reserved path", filepath.Clean(vm.MountPath)),
 				Paths:   []string{"mountPath"},
-			}).ViaFieldIndex("volumeMounts", i))
+			}).ViaIndex(i))
 		} else if !filepath.IsAbs(vm.MountPath) {
-			errs = errs.Also(apis.ErrInvalidValue(vm.MountPath, "mountPath").ViaFieldIndex("volumeMounts", i))
+			errs = errs.Also(apis.ErrInvalidValue(vm.MountPath, "mountPath").ViaIndex(i))
 		}
 		if !vm.ReadOnly {
-			errs = errs.Also(apis.ErrMissingField("readOnly").ViaFieldIndex("volumeMounts", i))
+			errs = errs.Also(apis.ErrMissingField("readOnly").ViaIndex(i))
 		}
 
-		if vm.SubPath != "" {
-			errs = errs.Also(
-				apis.ErrDisallowedFields("subPath").ViaFieldIndex("volumeMounts", i))
-		}
-		if vm.MountPropagation != nil {
-			errs = errs.Also(
-				apis.ErrDisallowedFields("mountPropagation").ViaFieldIndex("volumeMounts", i))
-		}
 	}
 
 	if missing := volumes.Difference(seen); missing.Len() > 0 {
 		errs = errs.Also(&apis.FieldError{
 			Message: fmt.Sprintf("volumes not mounted: %v", missing.List()),
-			Paths:   []string{"volumeMounts"},
+			Paths:   []string{""},
 		})
 	}
+	return errs
+}
 
-	// Some corev1.Container fields are set by Knative Serving controller.  We disallow them
-	// here to avoid silently overwriting these fields and causing confusions for
-	// the users.  See pkg/controller/revision/resources/deploy.go#makePodSpec.
-	var ignoredFields []string
-	if container.Name != "" {
-		ignoredFields = append(ignoredFields, "name")
-	}
-
-	if container.Lifecycle != nil {
-		ignoredFields = append(ignoredFields, "lifecycle")
-	}
-	if len(ignoredFields) > 0 {
-		// Complain about all ignored fields so that user can remove them all at once.
-		errs = errs.Also(apis.ErrDisallowedFields(ignoredFields...))
-	}
-	if err := validateContainerPorts(container.Ports); err != nil {
-		errs = errs.Also(err.ViaField("ports"))
-	}
-	// Validate our probes
-	if err := validateProbe(container.ReadinessProbe).ViaField("readinessProbe"); err != nil {
-		errs = errs.Also(err)
-	}
-	if err := validateProbe(container.LivenessProbe).ViaField("livenessProbe"); err != nil {
-		errs = errs.Also(err)
-	}
-	if _, err := name.ParseReference(container.Image, name.WeakValidation); err != nil {
-		fe := &apis.FieldError{
-			Message: "Failed to parse image reference",
-			Paths:   []string{"image"},
-			Details: fmt.Sprintf("image: %q, error: %v", container.Image, err),
+func validateDisallowedFields(request, maskedRequest interface{}) *apis.FieldError {
+	if disallowed, err := kmp.CompareSetFields(request, maskedRequest); err != nil {
```

Comment:
Hmm, this probably returns Go fields, not JSON fields, which isn't exactly what we want :(

Reply:
Maybe not, it looks like you fixed that in the test diffs?

Reply:
I fixed kmp to return JSON fields if they are present on the struct. :)